src/lib.rs

```diff
@@ -0,0 +1,489 @@
+lazy_static! {
+    static ref TRACES: RwLock<HashMap<JitLocation, JitTrace>> = RwLock::new(HashMap::new());
+}
+
+/// Execute the trace for the specified location (if there is one).
+///
+/// Returns true if a trace was present (and thus was executed), or false otherwise.
+fn try_exec_trace(loc: JitLocation) -> bool {
+    let trace_map = TRACES.read().unwrap();
+    let trace = trace_map.get(&loc);
+    if trace == Some(&JitTrace::FinalisedTrace) {
+        JitThreadContext::phase_transition(JitPhase::RunningTrace(loc));
+        // XXX Really execute the trace.
+        JitThreadContext::phase_transition(JitPhase::Interpreting);
+        return true;
+    }
+    false
+}
+
+/// Is a thread currently tracing/compiling the specified location?
+fn trace_pending(loc: JitLocation) -> bool {
+    let trace_map = TRACES.read().unwrap();
+    match trace_map.get(&loc) {
+        Some(&JitTrace::TracePending) => return true,
+        _ => return false,
+    };
+}
+
+/// Store a trace (or "trace pending" marker in the trace cache for the specified location.
+fn set_trace(loc: JitLocation, trace: JitTrace) {
+    // Sanity checks.
+    {
+        // Extra scope for dropping the lock prior to the write() lock later (avoids deadlock).
+        let trace_map = TRACES.read().unwrap();
+        let old_val = trace_map.get(&loc);
+        if trace == JitTrace::TracePending {
+            // If we are collecting a trace, then the location should not already be in the trace cache.
+            assert!(old_val == None);
+        } else if trace == JitTrace::FinalisedTrace {
+            // If we are storing a finished trace, then it should have been marked as "pending" first.
+            assert!(old_val == Some(&JitTrace::TracePending));
+        }
+    }
+
+    TRACES.write().unwrap().insert(loc, trace);
+}
+
+/// Shared hot counts.
+///
+/// Maps a location to its hot count. Increments of the counter are atomic due to the RwLock.
+lazy_static! {
+    static ref HOT_COUNTS: RwLock<HashMap<JitLocation, HotCount>> = RwLock::new(HashMap::new());
+}
+
+/// Get the hot count for the given location.
+fn hot_count(loc: JitLocation) -> HotCount {
+    let ret = match HOT_COUNTS.read().unwrap().get(&loc) {
+        Some(v) => *v,
+        None => 0,
+    };
+    ret
+}
+
+/// Atomically increment and return the hot count for the given location.
+fn incr_hot_count(loc: JitLocation) -> HotCount {
+    // XXX don't store locations which cannot be loop heads.
+    let mut counts = HOT_COUNTS.write().unwrap();
+    let loc_count = counts.entry(loc).or_insert(0);
+    *loc_count += 1;
+    *loc_count
+}
+
+/// Thread-specific meta-tracer context.
+///
+/// Each thread has one of instance of this inside a thread-local.
+#[derive(Debug)]
+pub struct JitThreadContext {
+    phase: JitPhase,
+    // XXX PT packet buffer goes here.
+}
+
+impl JitThreadContext {
+    pub fn new() -> Self {
```

Comment:
In an `impl` it's more common to write:

```
pub fn new() -> JitThreadContext {
    JitThreadContext {
        phase: ...
    }
}
```

because `Self` can only be `JitThreadContext`. [`Self` is probably best restricted to methods defined and used in traits.]

Reply:
I'm looking in the rustc sources, and I'm seeing a mix of both. I'll ask in IRC. Honestly, unless there are any gotchas I think `Self` is more maintainable, since it means you can rename the type in one place (i.e. `impl`).

Reply:
> oh yeah, one other question in a new(), is if preferred to build the value via the name `Self`, or as the verbatim type name?

> edd, it's a matter of taste, but I generally go for Self, and I think it's the convention now. certainly, clippy agrees.

If it's all the same to you, let's use `Self`.

Reply:
I'm not sure I feel strongly. If Clippy suggests that, let's go with it, since it's one less Clippy warning to ignore (and some of Clippy's warnings should definitely be ignored!).